### PR TITLE
Interactive UI for building queries

### LINKED ISSUE #10: UI for building queries
Provide an interactive way for users to build queries.

### PR DESCRIPTION
**Pull request by MatsDK**
closes #10 

This pull request changes the following:
- Updated UI, general layout didn't change
- Show user the ts type of the inputs for queries/mutations
- Query builder: users can insert queries/mutations with a generated default input
- Ability to set headers on requests 
- Store tabs and headers in localstorage

https://user-images.githubusercontent.com/67562518/180209132-c10eb107-f5cf-48b2-ba69-8987fae44bfc.mp4

### CODE REVIEW

**Comment by sachinraja** [comment]
Hey, thanks! This is *a lot*. I really wish you left a comment on the issue so we could've talked about this first. I'll try to review this within a few days.

**Comment by MatsDK** [comment]
Hey, thanks for taking a look at this and leaving some feedback.
At first, I was just experimenting with the package and didn't think about writing a comment on the issue.

The generation of the inputs is done by the server together with the typescript declarations.
When I get the time, I will try to remove the distinction between queries and mutations in the query builder.

File: packages/components/src/components/docs.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+const DocsForOperationType: React.FC<DocsForOperationTypeProps> = ({ label, operations }) => (
+  <div>
+    <h1 className='text-2xl font-semibold bg-primary p-3 sticky top-0 shadow-md'>
+      {label}
+      <span className="ml-2 text-gray-500 font-normal">{Object.keys(operations).length}</span>
+    </h1>
+    {Object.entries(operations).map(([name, def]) => (
+      <div className='py-2 px-3 border-b-2 border-primary'>
```
Comment (sachinraja) [inline packages/components/src/components/docs.tsx:44]: should have a `key` here I think

File: packages/components/src/components/queryBuilder/index.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          </button>
+        </div>
+        {queryBuilderOpen && (
+          <div className='flex-1 bg-primary w-full overflow-y-auto pt-2 px-4 pb-5'>
+            <span className="font-semibold">Operations</span>
+            <div className='pt-3'>
+              <div className='ml-1'>
+                {Object.keys({ ...types.queries, ...types.mutations }).map((opName, idx) => (
+                  <button
+                    key={idx}
```
Comment (sachinraja) [inline packages/components/src/components/queryBuilder/index.tsx:97]: key should not be index (this applies to everywhere)

File: packages/components/src/components/settings.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        <Header
+          removeHeader={() => {
+            const newHeaders = headerItems
+              .filter((_item, headerIdx) => idx !== headerIdx)
+              .reduce((prev, curr) => {
+                prev[curr[0]] = curr[1]
+                return prev
+              }, {} as HeadersType)
+
+            setHeaders(() => newHeaders)
```
Comment (sachinraja) [inline packages/components/src/components/settings.tsx:52]: use the `headers` passed as an argument to `setHeaders` instead of the atom to ensure consistency (this applies to all the methods here). Also don't compare indexes as those are unstable. Instead, compare something unique. Also use `splice` to remove something from an array.

File: packages/components/src/components/settings.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    <>
+      {headerItems.map(([name, value], idx) => (
+        <Header
+          removeHeader={() => {
+            const newHeaders = headerItems
+              .filter((_item, headerIdx) => idx !== headerIdx)
+              .reduce((prev, curr) => {
+                prev[curr[0]] = curr[1]
+                return prev
+              }, {} as HeadersType)
```
Comment (sachinraja) [inline packages/components/src/components/settings.tsx:50]: I don't get what this is for.

File: packages/components/src/components/sidebarOverlay.tsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+interface SidebarOverlayProps {
+  hide: () => void
+}
+
+export const SidebarOverlay: React.FC<SidebarOverlayProps> = ({ hide, children }) => {
+  return (
+    <div onClick={hide} className='absolute h-screen z-10 overflow-hidden bg-[#00000055] inset-0 w-full'>
+      <div
+        onClick={(e) => e.stopPropagation()}
+        className='absolute h-screen top-0 z-10 overflow-auto bg-secondary right-0 w-4/12	'
```
Comment (sachinraja) [inline packages/components/src/components/sidebarOverlay.tsx:12]: I see some `w-[fractionhere]`, please test to see if the UI scales across screen sizes.

File: packages/components/src/components/tab/group.tsx [hunk 6]
```diff
@@ -78,12 +78,12 @@ export const TabGroup = () => {
           onClick={() => {
             const newTab = createNewDefaultTab(totalTabsCreated)
             const newTabs: Tab[] = [...tabs, newTab]
-            setTabs(newTabs)
+            setTabs(() => newTabs)
             setTotalTabsCreated((totalTabsCreated) => totalTabsCreated + 1)
             updateCurrentTabId(newTab.id)
```
Comment (sachinraja) [inline packages/components/src/components/tab/group.tsx:83]: this is some really bad state management code I wrote here. But it's because I had a lot of trouble with getting the tabs to work (changing their positions and keeping their state intact). If you think you can rewrite it to make it better, then feel free to do that.

File: packages/components/src/components/tab/store.ts [hunk 7]
```diff
@@ -30,3 +68,6 @@ export const previousTabAtom = atom((get) => {
 export const currentTabIndexAtom = atom((get) => {
   return get(tabsAtom).findIndex((tab) => tab.id === get(currentTabIdAtom))
 })
+
+export const queryBuilderOpened = atom<boolean>(false)
```
Comment (sachinraja) [inline packages/components/src/components/tab/store.ts:72]: `<boolean>` shouldn't be necessary here

File: packages/components/src/components/tab/store.ts [hunk 8]
```diff
@@ -1,15 +1,53 @@
...
+    const oldValue = get(stateAtom)
+    const newValue = { ...oldValue, headers: updateHeaders(oldValue.headers) }
+
+    set(stateAtom, newValue)
+    localStorage.setItem('trpc-playground-config', JSON.stringify(newValue))
+  },
+)
+
+export const tabsAtom: WritableAtom<Tab[], (old: Tab[]) => Tab[]> = atom(
+  (get) => get(stateAtom).tabs || [],
```
Comment (sachinraja) [inline packages/components/src/components/tab/store.ts:41]: how can `tabs` be falsy here?

File: packages/components/src/components/toolbar.tsx [hunk 9]
```diff
@@ -1,11 +1,36 @@
-import { useAtom } from 'jotai'
+import { useAtom, atom } from 'jotai'
+import { Docs } from './docs'
 import { configAtom } from './provider'
+import { Settings } from "./settings"
+
+const showDocsAtom = atom(false),
+  showSettingsAtom = atom(false)
```
Comment (sachinraja) [inline packages/components/src/components/toolbar.tsx:7]: these should be two separate declarations

File: packages/components/src/utils/playground-request.ts [hunk 10]
```diff
@@ -27,5 +39,5 @@ export async function makePlaygroundRequest<Operation extends PlaygroundRequestO
     body: requestBody,
   })
 
-  return response.json()
+  return response.json() as Promise<GetTypesResponse>
```
Comment (sachinraja) [inline packages/components/src/utils/playground-request.ts:42]: this shouldn't be necessary

File: packages/html/postcss.config.cjs [hunk 11]
```diff
@@ -1,16 +1,18 @@
 module.exports = {
```
Comment (sachinraja) [inline packages/html/postcss.config.cjs:1]: please revert the style changes here

File: packages/trpc-playground/src/get-default-input.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+  ZodTypeDef,
+  ZodUnionDef,
+} from 'zod'
+
+const getDefaultForDef = (def: any): string | undefined => {
+  if (!def) return
+
+  switch ((def as any).typeName) {
+    case ZodFirstPartyTypeKind.ZodString:
+      return defaultString(def)
```
Comment (sachinraja) [inline packages/trpc-playground/src/get-default-input.ts:64]: what if the `typeName` is none of these cases?

File: packages/trpc-playground/src/get-default-input.ts [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+  return `null`
+}
+
+const defaultObject = (def: ZodObjectDef) => {
+  let ret = `{ `
+
+  let entries = Object.entries(def.shape())
+  entries.forEach(([name, propDef], idx) => {
+    ret += `${name}: ${getDefaultForDef(propDef._def)}`
+    if (idx != entries.length - 1) ret += `, `
```
Comment (sachinraja) [inline packages/trpc-playground/src/get-default-input.ts:147]: use `!==`

File: packages/trpc-playground/src/get-default-input.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+  })
+  ret += `}`
+
+  return ret
+}
+
+// TODO: Handle max/min length
+const defaultArray = (_def: ZodArrayDef) => {
+  return `[]`
+}
```
Comment (sachinraja) [inline packages/trpc-playground/src/get-default-input.ts:158]: arrays will usually have inner types

File: packages/trpc-playground/src/get-default-input.ts [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+const defaultSet = (_def: ZodSetDef) => {
+  return 'new Set()'
+}
+
+const defaultPromise = (def: ZodPromiseDef) => {
+  return `Promise.resolve(${getDefaultForDef(def.type._def)})`
+}
+
+// ZodLazy
```
Comment (sachinraja) [inline packages/trpc-playground/src/get-default-input.ts:224]: comment here should probably be removed

File: packages/trpc-playground/src/handlers/common.ts [hunk 16]
```diff
@@ -18,12 +18,12 @@ export const getCommonHandlerReqData = async (config: TrpcPlaygroundConfig) => {
...
   }
 
-  const types = await resolvedConfig.resolveTypes(config.router)
+  const types = await resolvedConfig.resolveTypes!(config.router)
 
   return {
     config: resolvedConfig,
     htmlPlaygroundPage,
-    stringifiedTypes: JSON.stringify(types),
+    types,
```
Comment (sachinraja) [inline packages/trpc-playground/src/handlers/common.ts:26]: Why? These are stringified here because they don't change, so we only need to call `JSON.stringify` once, not per-request

File: packages/trpc-playground/src/zod-resolve-types.ts [hunk 17]
```diff
@@ -28,12 +31,33 @@ const joinQueries = (functionName: string, queries: Record<string, { inputParser
...
-  joinQueries('query', router._def.queries),
-  joinQueries('mutation', router._def.mutations),
-]
+type DefaultOperationType = { value: string; inputLength: number }
+type QueryDefaultAndType = Record<string, { default: DefaultOperationType; type: string }>
+
+export type GetTypesFromRouterReturn = {
+  queries: QueryDefaultAndType
+  mutations: QueryDefaultAndType
+}
```
Comment (sachinraja) [inline packages/trpc-playground/src/zod-resolve-types.ts:40]: please extract these types to somewhere shared between the frontend and backend so they don't have to be duplicated

File: packages/trpc-playground/src/handlers/common.ts [hunk 18]
```diff
@@ -18,12 +18,12 @@ export const getCommonHandlerReqData = async (config: TrpcPlaygroundConfig) => {
     })
   }
 
-  const types = await resolvedConfig.resolveTypes(config.router)
+  const types = await resolvedConfig.resolveTypes!(config.router)
```
Comment (sachinraja) [inline packages/trpc-playground/src/handlers/common.ts:21]: `resolveTypes` can't be undefined, why is this here?

File: packages/typescript-extension/src/project.ts [hunk 19]
```diff
@@ -39,14 +39,19 @@ export class TypescriptProject {
...
+        system,
+        [TS_PROJECT_ENTRYPOINT],
+        typescript,
+        {
+          target: typescript.ScriptTarget.ESNext,
+        },
+      )
+    } catch (e) {
+      console.error(e)
+    }
```
Comment (sachinraja) [inline packages/typescript-extension/src/project.ts:54]: why?

File: packages/typescript-extension/src/tsfs.ts [hunk 20]
```diff
@@ -45,9 +45,11 @@ export class TSFS {
...
     if (isCached) {
       const fileNames = meta.files
-      const fileContents = (await Promise.all(
-        fileNames.map(f => localforage.getItem<string>(`ts-lib/${libName}/${meta.version}/${f}`)),
-      )) as string[] // type-cast is olay because we know this file should exist
+      const fileContents = (
+        await Promise.all(
+          fileNames.map(f => localforage.getItem<string>(`ts-lib/${libName}/${meta.version}/${f}`)),
+        )
+      ) as string[] // type-cast is olay because we know this file should exist
```
Comment (sachinraja) [inline packages/typescript-extension/src/tsfs.ts:52]: please don't unnecessarily reformat things. it makes it difficult to review

**Review by sachinraja — CHANGES_REQUESTED** [review]
Overall, this looks good, nice work! I really wish we could have split this up into multiple PRs to make it easy to review and incrementally improve on. It's difficult to merge all of this at once, especially when I don't understand some of it.

It would help a lot to provide an overview of all the changes you made and why you made them.

File: packages/components/src/components/docs.tsx (hunk 1 above)
Comment (MatsDK) [reply packages/components/src/components/docs.tsx:44]: done

File: packages/components/src/components/tab/group.tsx (hunk 6 above)
Comment (MatsDK) [reply packages/components/src/components/tab/group.tsx:83]: I also found some issues with the tabs state, when creating a new tab there sometimes occurs a collision of id's.
I think this can we solved by storing the `totalTabsCreated` in localStorage or we could get rid of `totalTabsCreated` and just create a random id when a new tab is created.

What do you think?

File: packages/components/src/components/settings.tsx (hunk 4 above)
Comment (MatsDK) [reply packages/components/src/components/settings.tsx]: removed

File: packages/components/src/components/settings.tsx (hunk 3 above)
Comment (MatsDK) [reply packages/components/src/components/settings.tsx]: removed

File: packages/components/src/components/tab/store.ts (hunk 8 above)
Comment (MatsDK) [reply packages/components/src/components/tab/store.ts]: It can't be false -> removed

File: packages/components/src/components/toolbar.tsx (hunk 9 above)
Comment (MatsDK) [reply packages/components/src/components/toolbar.tsx]: done

File: packages/trpc-playground/src/get-default-input.ts (hunk 12 above)
Comment (MatsDK) [reply packages/trpc-playground/src/get-default-input.ts:64]: Maybe we could just return an empty string?

File: packages/trpc-playground/src/get-default-input.ts (hunk 14 above)
Comment (MatsDK) [reply packages/trpc-playground/src/get-default-input.ts]: added default for inner type.

File: packages/trpc-playground/src/handlers/common.ts (hunk 16 above)
Comment (MatsDK) [reply packages/trpc-playground/src/handlers/common.ts]: Yes, that makes sense.
fixed

File: packages/trpc-playground/src/zod-resolve-types.ts (hunk 17 above)
Comment (MatsDK) [reply packages/trpc-playground/src/zod-resolve-types.ts]: Moved it inside `@trpc-playground/types`.

File: packages/trpc-playground/src/handlers/common.ts (hunk 18 above)
Comment (MatsDK) [reply packages/trpc-playground/src/handlers/common.ts]: I don't know why I did that, removed it.

File: packages/typescript-extension/src/project.ts (hunk 19 above)
Comment (MatsDK) [reply packages/typescript-extension/src/project.ts]: try-catch removed

File: packages/trpc-playground/src/get-default-input.ts (hunk 15 above)
Comment (MatsDK) [reply packages/trpc-playground/src/get-default-input.ts]: done

File: packages/components/src/components/tab/group.tsx (hunk 6 above)
Comment (sachinraja) [reply packages/components/src/components/tab/group.tsx:83]: I think the random id is a good solution for now

File: packages/trpc-playground/src/get-default-input.ts (hunk 12 above)
Comment (sachinraja) [reply packages/trpc-playground/src/get-default-input.ts:64]: sounds good to me, I think this is what was already happening actually

**Comment by sachinraja** [comment]
Looks mostly good now, I'll do one last pass over the changed code + test locally, and then merge (probably not today).

**Comment by MatsDK** [comment]
Thanks for taking a look at this. (and sorry that this is so much)
I just created a commit that resolves the issue with the tab ids.

File: packages/components/package-lock.json [hunk 21]
```diff
@@ -0,0 +1,1665 @@
+{
```
Comment (sachinraja) [inline packages/components/package-lock.json:1]: This file should be removed, this project uses pnpm

File: packages/components/src/components/editor.tsx [hunk 22]
```diff
@@ -29,22 +38,23 @@ export const Editor = () => {
   const [responseValue, setResponseValue] = useAtom(responseValueAtom)
   const [trpcClient] = useAtom(trpcClientAtom)
   const [config] = useAtom(configAtom)
-  const evalFunction = useMemo(() => config.request.batching ? batchEval : serialEval, [config.request.batching])
+  const evalFunction = useMemo(() => config?.request.batching ? batchEval : serialEval, [config.request.batching])
```
Comment (sachinraja) [inline packages/components/src/components/editor.tsx:41]: `config` should not be optional

File: packages/components/src/components/editor.tsx [hunk 23]
```diff
@@ -29,22 +38,23 @@ export const Editor = () => {
...
     try {
-      const types = await makePlaygroundRequest('getTypes', {
+      const response = await makePlaygroundRequest('getTypes', {
         playgroundEndpoint: config.playgroundEndpoint,
       })
+      setTypes(response)
 
       editorView.dispatch(injectTypes({
-        '/index.d.ts': types.join('\n'),
+        '/index.d.ts': response.tsTypes.join('\n'),
```
Comment (sachinraja) [inline packages/components/src/components/editor.tsx:53]: Maybe we should change the name of the operation from `getTypes`? Maybe to `getRouterSchema` instead.

File: packages/components/src/components/queryBuilder/index.tsx [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+
+export const QueryBuilder: React.FC = () => {
+  const [types] = useAtom(typesAtom)
+  const [queryBuilderOpen, setQueryBuilderOpened] = useAtom(queryBuilderOpened)
+  const [generated] = useAtom(generatedAtom)
+  const [operationName, setOperationName] = useAtom(operationNameAtom)
+  const [editorView] = useAtom(editorAtom)
+
+  // On insert btn click
+  const insertGenerated = useCallback(
```
Comment (sachinraja) [inline packages/components/src/components/queryBuilder/index.tsx:37]: Can we get it to insert at the current cursor position?

File: packages/components/src/components/queryBuilder/index.tsx [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+      // Add generated query to next line in editor
+      editorView.dispatch({
+        changes: {
+          from: line.from,
+          to: line.to,
+          insert: `${line.text}\n${generated}`,
+        },
+      })
+
+      // Select generated default input so user can instantly remove/edit
```
Comment (sachinraja) [inline packages/components/src/components/queryBuilder/index.tsx:56]: This doesn't seem to work, it's not selected for me

File: packages/components/src/components/settings.tsx [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+        className='mt-1 bg-primary border-zinc-800 border px-2 text-lg outline-none text-zinc-200 py-1'
+        onClick={() => {
+          if ("name" in headers)
+            return alert("Header with the name `name` already exists")
+
+          setHeaders((headers: HeadersType) => {
+            const newHeaders = { ...headers }
+            newHeaders["name"] = ""
+            return newHeaders
+          })
```
Comment (sachinraja) [inline packages/components/src/components/settings.tsx:62]: `"name"` is hardcoded here? This seems wrong

File: packages/components/src/components/settings.tsx [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+        size={name.length}
+        type="text"
+        defaultValue={name}
+        className='bg-primary border-zinc-800 border px-2 text-lg outline-none min-w-[130px] text-zinc-200 py-1 h-9'
+        readOnly
+        onClick={() => {
+          const newName = prompt("Rename header", name)?.trim()
+          if (newName == null || newName === name || !newName) return
+
+          if (newName in headers) return alert("Name already exists")
```
Comment (sachinraja) [inline packages/components/src/components/settings.tsx:91]: `"Name"` is hardcoded here too

File: packages/components/src/components/tab/group.tsx [hunk 28]
```diff
@@ -76,14 +75,13 @@ export const TabGroup = () => {
         <button
           title='Create new tab'
           onClick={() => {
-            const newTab = createNewDefaultTab(totalTabsCreated)
+            const newTab = createNewDefaultTab()
             const newTabs: Tab[] = [...tabs, newTab]
-            setTabs(newTabs)
-            setTotalTabsCreated((totalTabsCreated) => totalTabsCreated + 1)
+            setTabs(() => newTabs)
             updateCurrentTabId(newTab.id)
```
Comment (sachinraja) [inline packages/components/src/components/tab/group.tsx:81]: Thanks, this looks nicer now

File: packages/html/index.html [hunk 29]
```diff
@@ -1,12 +1,15 @@
 <!DOCTYPE html>
```
Comment (sachinraja) [inline packages/html/index.html:1]: formatting in this file should be reverted

File: packages/trpc-playground/src/get-default-input.ts [hunk 30]
```diff
@@ -0,0 +1,221 @@
...
+    else ret += ` `
+  })
+  ret += `}`
+
+  return ret
+}
+
+const defaultArray = (def: ZodArrayDef) => {
+  return `[${getDefaultForDef(def.type._def)}]`
+}
```
Comment (sachinraja) [inline packages/trpc-playground/src/get-default-input.ts:157]: nice

File: packages/types/src/index.d.ts [hunk 31]
```diff
@@ -63,7 +63,7 @@ export type ServerConfig = {
   /**
    * Resolves the typescript types for the router and returns an array of types to inject. Uses [`zod-to-ts`](https://github.com/sachinraja/zod-to-ts) by default.
    */
-  resolveTypes?: (router: AnyRouter) => string[] | Promise<string[]>
+  resolveTypes?: (router: AnyRouter) => ResolveTypesReturn | Promise<ResolveTypesReturn>
```
Comment (sachinraja) [inline packages/types/src/index.d.ts:66]: Where is the `ResolveTypesReturn` type coming from? I don't see it in this file

**Review by sachinraja — CHANGES_REQUESTED** [review]
Looks great! Thanks for sticking with this. Just a few more changes and this is good.

Also noticed a few formatting issues and the workflow is failing. Try running `pnpm format` in the root directory to fix that.

File: packages/types/src/index.d.ts (hunk 31 above)
Comment (MatsDK) [reply packages/types/src/index.d.ts:66]: Should `ResolveTypesReturn` be renamed to something like `ResolvedRouterSchema` since the operation name changed to `getRouterSchema`?

File: packages/components/src/components/queryBuilder/index.tsx (hunk 25 above)
Comment (MatsDK) [reply packages/components/src/components/queryBuilder/index.tsx:56]: That is weird, for me this is happening

image

File: packages/components/src/components/settings.tsx (hunk 26 above)
Comment (MatsDK) [reply packages/components/src/components/settings.tsx:62]: fixed

File: packages/components/src/components/settings.tsx (hunk 27 above)
Comment (MatsDK) [reply packages/components/src/components/settings.tsx:91]: fixed

File: packages/types/src/index.d.ts (hunk 31 above)
Comment (sachinraja) [reply packages/types/src/index.d.ts:66]: Yes but I also don't see it in this file so I don't know where it's being imported from

File: packages/components/src/components/queryBuilder/index.tsx (hunk 25 above)
Comment (sachinraja) [reply packages/components/src/components/queryBuilder/index.tsx:56]: Hmm I'll test it again.

File: packages/components/src/components/editor.tsx (hunk 23 above)
Comment (MatsDK) [reply packages/components/src/components/editor.tsx:53]: done

File: packages/types/src/index.d.ts (hunk 31 above)
Comment (MatsDK) [reply packages/types/src/index.d.ts]: Should be fixed now

File: packages/components/src/components/sidebarOverlay.tsx (hunk 5 above)
Comment (MatsDK) [reply packages/components/src/components/sidebarOverlay.tsx]: I haven't found any issues with this

**Comment by sachinraja** [comment]
Ah I forgot about this! So sorry about that. Will merge and publish today.

**Comment by n-glaz** [comment]
Looking forward to trying this out! :)

**Review by sachinraja — APPROVED** [review]
Awesome work!

**Comment by sachinraja** [comment]
@MatsDK If you can get around to it, a video showing off the new playground (like what we currently have on the README) would be great. Otherwise, I can get to it later.

**Comment by sachinraja** [comment]
@MatsDK @n-glaz this is now published, update `trpc-playground` to `0.2.0`.

**Comment by MatsDK** [comment]
Thanks for merging this PR.

I think the issue with the tabs is that we are only setting the tabs state in the `useLayoutEffect` when you switch tabs.

https://github.com/sachinraja/trpc-playground/blob/2ba7d62272c62b82b4334db4bf3bf6621eacb37c/packages/components/src/components/editor.tsx#L98-L108

Maybe we should listen for `onUpdate` on the editor, what do you think?
Should I create an issue for this?

**Comment by sachinraja** [comment]
I chose that way because setting the state on every input seems inefficient. It would re-render the editor on every change.

**Comment by MatsDK** [comment]
Oh yes, how can we make sure that changes get stored inside localstorage and do not get lost when the paged is refreshed or something? Maybe something similar like the `refreshTypes` polling with an interval?

**Comment by sachinraja** [comment]
I think we can just hook into the `onbeforeunload` event
